src/librustc_mir/build/matches/mod.rs

```diff
@@ -793,47 +924,130 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
+        // candidate.
+        let re_empty = self.hir.tcx().types.re_empty;
+        for binding in bindings {
+            let source_info = self.source_info(binding.span);
+            let local_for_guard = self.storage_live_binding(
+                block, binding.var_id, binding.span, WithinGuard);
+            // Question: Why schedule drops if bindings are all
+            // shared-&'s?  Answer: Because schedule_drop_for_binding
+            // also emits StorageDead's for those locals.
+            self.schedule_drop_for_binding(binding.var_id, binding.span, WithinGuard);
+            match binding.binding_mode {
+                BindingMode::ByValue => {
+                    let rvalue = Rvalue::Ref(re_empty, BorrowKind::Shared, binding.source.clone());
+                    self.cfg.push_assign(block, source_info, &local_for_guard, rvalue);
+                }
+                BindingMode::ByRef(region, borrow_kind) => {
+                    // NOTE tricky business: For `ref id` and `ref mut
+                    // id` patterns, we want `id` within the guard to
+                    // correspond to a temp of type `& &T` or `& &mut
+                    // T`, while within the arm body it will
+                    // correspond to a temp of type `&T` or `&mut T`,
+                    // as usual.
+                    //
+                    // But to inject the level of indirection, we need
+                    // something to point to.
+                    //
+                    // So:
+                    //
+                    // 1. First set up the local for the arm body
+                    //   (even though we have not yet evaluated the
+                    //   guard itself),
+                    //
+                    // 2. Then setup the local for the guard, which is
+                    //    just a reference to the local from step 1.
+                    //
+                    // Note that since we are setting up the local for
+                    // the arm body a bit eagerly here (and likewise
+                    // scheduling its drop code), we should *not* do
+                    // it redundantly later on.
+                    //
+                    // While we could have kept track of this with a
+                    // flag or collection of such bindings, the
+                    // treatment of all of these cases is uniform, so
+                    // we should be safe just avoiding the code
+                    // without maintaining such state.)
+                    let local_for_arm_body = self.storage_live_binding(
+                        block, binding.var_id, binding.span, OutsideGuard);
+                    self.schedule_drop_for_binding(binding.var_id, binding.span, OutsideGuard);
+                    let rvalue = Rvalue::Ref(region, borrow_kind, binding.source.clone());
+                    self.cfg.push_assign(block, source_info, &local_for_arm_body, rvalue);
```

Comment:
So before I thought we were going to need some sort of cast here -- I was trying to remember why -- I believe it was looking forward to the point where we try to maintain a shared borrow the value being matched over the whole match, in which case this first `&mut` borrow would be illegal...

Reply:
okay let me look into this, see if I can make a test case that exposes the scenario you describe.

Reply:
> I believe it was looking forward to the point where we try to maintain a shared borrow the value being matched over the whole match, in which case this first `&mut` borrow would be illegal...

Isn't the answer here is "this is why we have 2-phase borrows"? There's this cute example, which compiles today, is not a soundness issue, and works in the 2phi-borrow interpretation, but does not work if 2phi is not used correctly.

```Rust
use std::cell::Cell;

fn main() {
    let mut b = 4;
    let a : Cell<Option<&u32>> = Cell::new(None);
    match b {
        ref mut btag if { a.set(Some(&*btag)); true } => {
            // `btag` is immutably borrowed here.
            println!("a={:?} b={}", a, btag);
        }
        _ => {}
    }
}
```